## Analyze to understand the following question:
## How does physical activity relate to sleep and calories burned?

In [46]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("mysql+pymysql://root:root@127.0.0.1:3307/fitbit_db")


Average number of steps per day

In [47]:
query = """
SELECT 
    AVG(total_steps) AS avg_daily_steps
FROM daily_activity;
"""

df = pd.read_sql(query,engine)
print(df.head())

   avg_daily_steps
0        7637.9106


What times of day are there the most activities?

In [48]:
query = """
SELECT 
    hour,
    SUM(steps) AS total_steps
FROM hourly_steps
GROUP BY hour
ORDER BY total_steps DESC;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

   hour  total_steps
0    18     542848.0
1    19     528552.0
2    12     505848.0
3    17     498511.0
4    14     497813.0


Average hours of sleep

In [49]:
query = """
SELECT 
    AVG(minutes_asleep)/60 AS avg_sleep_hours
FROM sleep;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())


   avg_sleep_hours
0          6.98622


Relationship between steps and calories

In [50]:
query = """
SELECT 
    total_steps,
    calories
FROM daily_activity
ORDER BY total_steps DESC
LIMIT 20;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())


   total_steps  calories
0        36019      2690
1        29326      4547
2        27745      4398
3        23629      3808
4        23186      3921


Are more steps per day linked to more sleep?

In [51]:
query = """
SELECT 
    d.activity_date,
    d.total_steps,
    s.minutes_asleep
FROM daily_activity d
JOIN sleep s
ON d.user_id = s.user_id
AND d.activity_date = DATE(s.sleep_date);
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

  activity_date  total_steps  minutes_asleep
0    2016-04-12        13162             327
1    2016-04-13        10735             384
2    2016-04-15         9762             412
3    2016-04-16        12669             340
4    2016-04-17         9705             700


Average sleep by activity level

In [52]:
query = """
SELECT 
    CASE
        WHEN total_steps < 5000 THEN 'Low Activity'
        WHEN total_steps BETWEEN 5000 AND 10000 THEN 'Medium Activity'
        ELSE 'High Activity'
    END AS activity_level,
    AVG(minutes_asleep)/60 AS avg_sleep_hours
FROM daily_activity d
JOIN sleep s
ON d.user_id = s.user_id
AND d.activity_date = DATE(s.sleep_date)
GROUP BY activity_level;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())


    activity_level  avg_sleep_hours
0    High Activity         6.601010
1  Medium Activity         7.035682
2     Low Activity         7.571528


Average heart rate by hour per day

In [53]:
query = """
SELECT 
    HOUR(timestamp) AS hour,
    AVG(heartrate) AS avg_heartrate
FROM heartrate_seconds
GROUP BY hour
ORDER BY hour;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

   hour  avg_heartrate
0     0        66.7236
1     1        65.6424
2     2        63.5665
3     3        61.1435
4     4        60.2318


Correlation between steps and calories by user

In [54]:
query = """
SELECT user_id, AVG(total_steps) AS avg_steps, AVG(calories) AS avg_calories
FROM daily_activity
GROUP BY user_id;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

      user_id   avg_steps  avg_calories
0  1503960366  12116.7419     1816.4194
1  1624580081   5743.9032     1483.3548
2  1644430081   7282.9667     2811.3000
3  1844505072   2580.0645     1573.4839
4  1927972279    916.1290     2172.8065


Do days with good sleep (7+ hours) = more steps?

In [55]:
query = """
SELECT 
    CASE WHEN minutes_asleep >= 420 THEN 'Good Sleep' ELSE 'Poor Sleep' END AS sleep_quality,
    AVG(d.total_steps) AS avg_steps,
    AVG(d.calories) AS avg_calories
FROM daily_activity d
JOIN sleep s ON d.user_id = s.user_id AND d.activity_date = DATE(s.sleep_date)
GROUP BY sleep_quality;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

  sleep_quality  avg_steps  avg_calories
0    Poor Sleep  9362.9945     2414.1602
1    Good Sleep  7844.5895     2369.6419


Time in bed without sleep (anxiety/insomnia)

In [56]:
query = """
SELECT AVG(minutes_awake_in_bed) AS avg_awake_in_bed
FROM sleep;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

   avg_awake_in_bed
0           39.3098


Weekly pattern - What day of the week do you have the most steps?

In [57]:
query = """
SELECT DAYNAME(activity_date) AS day_name, AVG(total_steps) AS avg_steps
FROM daily_activity
GROUP BY day_name
ORDER BY avg_steps DESC;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

    day_name  avg_steps
0   Saturday  8152.9758
1    Tuesday  8125.0066
2     Monday  7780.8667
3  Wednesday  7559.3733
4     Friday  7448.2302


Peak hours of high heart rate (related to physical activity)

In [58]:
query = """
SELECT HOUR(timestamp) AS hour, AVG(heartrate) AS avg_hr
FROM heartrate_seconds
WHERE heartrate > 100
GROUP BY hour
ORDER BY avg_hr DESC;
"""

df2 = pd.read_sql(query, engine)
print(df2.head())

   hour    avg_hr
0    16  123.0331
1     2  122.6662
2    18  121.6934
3    17  121.4526
4    12  121.1925


## Key Findings: Physical Activity, Sleep & Calories

---

### General Activity
- Average of **7,638 steps/day** — slightly below the recommended 10,000
- Most active days: **Saturday & Tuesday**
- Peak activity hours: **6–7 PM** (after work) and **12 PM** (lunch break)

---

### Activity & Sleep — Surprising Finding
The data reveals a **counter-intuitive pattern**:

| Activity Level | Avg Sleep Hours |
|---|---|
| Low Activity (<5k steps) | **7.57 hrs** |
| Medium Activity (5k–10k) | **7.04 hrs** |
| High Activity (>10k steps) | **6.60 hrs** |

- Days with **poor sleep** (<7 hrs) → avg **9,363 steps**
- Days with **good sleep** (7+ hrs) → avg **7,845 steps**

>  **Interpretation:** Highly active people sleep less — likely due to busy schedules, not because activity harms sleep quality.

---

###  Activity & Calories
- Clear positive relationship: **more steps = more calories burned**
- High variability between users suggests **body composition** plays a significant role

---

###  Heart Rate Patterns
- Peak elevated heart rate at **4 PM** — aligns with the afternoon activity build-up before the 6 PM steps peak
- Unusual spike at **2 AM** — possible outliers or night-shift users

---

###  Sleep Quality
- Users spend an average of **39 minutes awake in bed**
- Indicates **poor sleep quality** for a significant portion of users

---

###  Bottom Line
> Higher physical activity **does** correlate with more calories burned,  
> but does **not** lead to better sleep.  
> Active users sleep *less* — suggesting a **busy lifestyle** drives both high activity and shorter sleep.